### PR TITLE
[FIX] point_of_sale: fix bug missing order in BackOffice:

### PR DESCRIPTION
**Pull request by anothingguy**
- Mutex of push_order should return a Promise to make push_order once at time
- Should be deep copy "orders" that need to "_save_to_server" to avoid deleting wrong orders which added to localStorage while "_save_to_server" is in progress.

Description of the issue/feature this PR addresses:

Current behavior before PR:
1. Turn off internet and make many orders (e.g 5 orders) in offline mode on POS
2. Turn on internet again, then click button "synch" orders on top-right on screen to push to server.
3. While step "2" is processing, try to make another order then validate this one immediately when step "2" doesn't finish yet.

Result:
- step 3 can run while step 2 is not finished yet => Mutex in "push_order" doesn't work properly
- Order in step 3 doesn't exist in Back-end Office because removal of orders of Step 2 when pushing in step 2 successfull.

Desired behavior after PR is merged:
- Mutex in "push_order" should work properly.
- Orders in localStorage shoud be removed correctly.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr
PR of my CLA: https://github.com/odoo/odoo/pull/13971

### CODE REVIEW

File: addons/point_of_sale/static/src/js/models.js [hunk 1]
```diff
@@ -746,7 +748,11 @@ exports.PosModel = Backbone.Model.extend({
         var self = this;
         this.set('synch',{ state: 'connecting', pending: orders.length});
 
-        return self._save_to_server(orders, options).done(function (server_ids) {
+        // Need to deep copy orders to avoid deleting new orders (which are saved to "orders" in localStorage also)
+        // while previous orders are in progress of "_save_to_order" and removed them after "_save_to_order" successfull
+        var flush_orders = JSON.parse(JSON.stringify(orders));
```
Comment (jorenvo) [inline addons/point_of_sale/static/src/js/models.js]: I agree with the fact that there's an issue here, but I'm not sure yet we want to create a copy of the orders. Technically it's a relatively expensive operation and in the end we don't really need an entire copy except for the ids. Conceptually I'm not sure I like the idea. An issue like this should be solved by either some mutex or some other mechanism (like eg. checking what orders exactly have been synced to the server instead of deleting them all). I'll think about it further in the next coming days.

**Comment by jorenvo** [comment]
@anothingguy thank you for the detailed explanation of the issue and the accompanying pull request. It's much appreciated!

@mart-e thanks for pinging me. There is definitely an issue here, I took some time to verify that. I'm just not 100% sure yet on the correct way to solve it yet, I'll come back to this in the next days.

File: addons/point_of_sale/static/src/js/models.js (hunk 1 above)
Comment (anothingguy) [inline addons/point_of_sale/static/src/js/models.js]: @jorenvo: thanks for your comments above, i total agree all. I will try to improve this too :)

Comment (anothingguy) [inline addons/point_of_sale/static/src/js/models.js]: @jorenvo : after re-thinking about this, i think we should copy `orders` than writing `some mutex or some other checking` as you suggest above to respect the consistency of `orders`. 
I think `orders` should be the same before and after processing `_save_to_server`. So it helps us avoid any traps if `_save_to_server` is inherited somewhere in another module and use `orders` data to do something later.
About technically, i agree deep copy is relatively expensive operation, so i switched to shallow copy to save performance, i tested and shallow copy is enough in this case. Base on some test http://jsben.ch/#/jCucv, i chose `ES6 Object.assign()` to copy `orders`.

Waiting for your idea.
Regards

File: addons/point_of_sale/static/src/js/models.js [hunk 2]
```diff
@@ -746,7 +748,11 @@ exports.PosModel = Backbone.Model.extend({
         var self = this;
         this.set('synch',{ state: 'connecting', pending: orders.length});
 
-        return self._save_to_server(orders, options).done(function (server_ids) {
+        // Need to shallow copy orders to avoid deleting new orders (which are saved to "orders" in localStorage also)
+        // while previous orders are in progress of "_save_to_order" and removed them after "_save_to_order" successfull
+        var flush_orders = Object.assign([], orders);
+
+        return self._save_to_server(flush_orders, options).done(function (server_ids) {
```
Comment (jorenvo) [inline addons/point_of_sale/static/src/js/models.js]: Ideally I think the server should return references to the orders that have actually been saved. It kind of does this, but it returns the pos.order record IDs which isn't particularly useful because we don't know what database ID corresponds with what JSON order. It would have been nicer if it returned instead the ```name``` property of the synced JSON orders. Then we could simply use that in the callback of ```posOrderModel.call('create_from_ui')``` to only delete orders actually synced. We cannot make such a change in a stable release though. In theory we could do another RPC call in JS to read the ```pos_reference``` (= JSON ```name```) fields of those pos.orders. But I like to avoid additional RPC if possible.

So instead how about the following:

Before the RPC call to ```create_from_ui``` we save the ```id``` properties of the objects in the ```orders``` array:

```javascript
var order_ids = _.pluck(orders, 'id');
```

After this we use that array in the callback when deleting orders

```javascript
_.each(order_ids, function (id) {
    self.db.remove_order(id);
});
```

I think this should resolve the problem since I think that the issue occurs in the time between the ```create_from_ui``` call and the callback right? Don't hesitate to point out if I'm missing something :).

**Comment by anothingguy** [comment]
@jorenvo: 
- I'm not sure it's better to `return the name property of synched JSON orders` when `create_from_ui` have done. Because sometime, i really need to know the ids of created orders in server to update some datas in POS front-end. I think `create` function should return `ids` as usually.
- About saving list `order_ids` to use in the callback when deleting orders, i already thought about it, it's really a simply away to solve to problem. But problem of consistency of `orders` is still there. 
For example, in my `pos_voucher` module, i have a `voucher` attached to `order`, when `order` pushed to server and make changes to vouchers of related order (e.g: update customer who used voucher, state, used_date of), than i need to reload vouchers to get latest info of voucher in POS front-end. 
```
        _save_to_server: function (orders, options) {
            var self = this;
            return _super._save_to_server.apply(this, arguments)
            .then(function(server_ids){
                // reload vouchers if order sell/use vouchers
                var voucher_ids = []
                _.each(orders, function(order){
                    var data = order && order.data || {};
                    voucher_ids = _.union(data.vouchers || []);
                })
                return self.reload_vouchers(voucher_ids)
                    .then(function(){
                        return server_ids;
                })
            })
        }
```
Because number of `orders` before and after running `_save_to_server` is not the same, so it's not 100% correct and maybe can raise some unexpected issue.
So from my point of view: the root cause comes from `orders` is reference variable of `cache` in `db.js`. So i really want to shallow copy it with lowest cost.

To sum up, i will make a change as you request as first simple step to fix issue. Odoo team can go futher to improve it in future if needed.

Regards.

**Comment by jorenvo** [comment]
Merged in 9.0 in dff352acdb8b15b59bb4257814e62c4e46798739 . Thanks for your work in investigating  and fixing this issue!

Your concerns about developing things that have to somehow figure out what orders have been synced are somewhat separate from the initial issue. So I prefer to merge and close this. Although it's not pretty, when overriding ```_save_to_server``` you could do a similar thing, and save the order ids before calling super I think.
